### ISSUE TITLE
Bertopic soemtimes gives me topic that are not mutually exclusive..

### ISSUE THREAD

**Issue by zhimin-z** [MaartenGr/BERTopic#1419]
![image](https://github.com/MaartenGr/BERTopic/assets/8592144/21cdefc2-3069-4a4e-9faf-222fd2ff611f)
As seen from above, topic 7 is included in topic 0, which makes the topic interpretation confusing. 
I have no idea what I can do with it. @MaartenGr Any suggestion for hyperparameter tuning?

```
from sklearn.feature_extraction.text import TfidfVectorizer
from gensim.models.coherencemodel import CoherenceModel
from bertopic.vectorizers import ClassTfidfTransformer
from sentence_transformers import SentenceTransformer
from bertopic.representation import KeyBERTInspired
from bertopic import BERTopic
from hdbscan import HDBSCAN
from umap import UMAP

import gensim.corpora as corpora
import pandas as pd
import wandb
import os

path_output = os.path.join(os.getcwd(), 'Result', 'RQ1', 'Special Topics')
path_model = os.path.join(os.getcwd(), 'Code', 'RQ1', 'Special Topic Modeling', 'Model')
if not os.path.exists(path_model):
    os.makedirs(path_model)

wandb_project = 'asset-management-topic-modeling'

os.environ["WANDB_API_KEY"] = XXXXX
os.environ["TOKENIZERS_PARALLELISM"] = "true"
os.environ["WANDB__SERVICE_WAIT"] = "100"

# set default sweep configuration
config_defaults = {
    # Refer to https://www.sbert.net/docs/pretrained_models.html
    'model_name': 'all-mpnet-base-v2',
    'metric_distane': 'manhattan',
    'calculate_probabilities': True,
    'reduce_frequent_words': True,
    'prediction_data': True,
    'low_memory': False,
    'random_state': 42,
    'ngram_range': 2,
}

config_sweep = {
    'method': 'grid',
    'metric': {
        'name': 'Coherence CV',
        'goal': 'maximize'
    },
    'parameters': {
        'n_components': {
            'values': [3, 4, 5, 6, 7],
        },
    }
}


class TopicModeling:
    def __init__(self, topic_type, min_cluster_size=20):
        # Initialize an empty list to store top models
        self.top_models = []
        self.path_model = path_model
        
        df = pd.read_json(os.path.join(path_output, 'preprocessed.json'))
        if topic_type == 'anomaly':
            df = df[df['Challenge_type'] == 'anomaly']
            self.docs = df[df['Challenge_summary'] != 'na']['Challenge_summary'].tolist() + df[df['Challenge_root_cause'] != 'na']['Challenge_root_cause'].tolist()
        elif topic_type == 'solution':
            self.docs = df[df['Solution'] != 'na']['Solution'].tolist()
        
        config_defaults['min_cluster_size'] = min_cluster_size
        config_sweep['name'] = topic_type
        config_sweep['parameters']['min_samples'] = {
            'values': list(range(1, config_defaults['min_cluster_size'] + 1))
        }
        
    def __train(self):
        # Initialize a new wandb run
        with wandb.init() as run:
            # update any values not set by sweep
            run.config.setdefaults(config_defaults)

            # Step 1 - Extract embeddings
            embedding_model = SentenceTransformer(run.config.model_name)

            # Step 2 - Reduce dimensionality
            umap_model = UMAP(n_components=wandb.config.n_components, metric=run.config.metric_distane,
                              random_state=run.config.random_state, low_memory=run.config.low_memory)

            # Step 3 - Cluster reduced embeddings
            hdbscan_model = HDBSCAN(min_cluster_size=run.config.min_cluster_size,
                                    min_samples=wandb.config.min_samples, prediction_data=run.config.prediction_data)

            # Step 4 - Tokenize topics
            vectorizer_model = TfidfVectorizer(ngram_range=(1, run.config.ngram_range))

            # Step 5 - Create topic representation
            ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=run.config.reduce_frequent_words)

            # Step 6 - Fine-tune topic representation
            representation_model = KeyBERTInspired()

            # All steps together
            topic_model = BERTopic(
                embedding_model=embedding_model,
                umap_model=umap_model,
                hdbscan_model=hdbscan_model,
                vectorizer_model=vectorizer_model,
                ctfidf_model=ctfidf_model,
                representation_model=representation_model,
                calculate_probabilities=run.config.calculate_probabilities
            )

            topics, _ = topic_model.fit_transform(self.docs)

            # Preprocess Documents
            documents = pd.DataFrame({"Document": self.docs,
                                      "ID": range(len(self.docs)),
                                      "Topic": topics})
            documents_per_topic = documents.groupby(
                ['Topic'], as_index=False).agg({'Document': ' '.join})
            cleaned_docs = topic_model._preprocess_text(
                documents_per_topic.Document.values)

            # Extract vectorizer and analyzer from BERTopic
            vectorizer = topic_model.vectorizer_model
            analyzer = vectorizer.build_analyzer()

            # Extract features for Topic Coherence evaluation
            tokens = [analyzer(doc) for doc in cleaned_docs]
            dictionary = corpora.Dictionary(tokens)
            corpus = [dictionary.doc2bow(token) for token in tokens]
            topic_words = [[words for words, _ in topic_model.get_topic(
                topic)] for topic in range(len(set(topics))-1)]

            coherence_cv = CoherenceModel(
                topics=topic_words,
                texts=tokens,
                corpus=corpus,
                dictionary=dictionary,
                coherence='c_v'
            )

            coherence_umass = CoherenceModel(
                topics=topic_words,
                texts=tokens,
                corpus=corpus,
                dictionary=dictionary,
                coherence='u_mass'
            )

            coherence_cuci = CoherenceModel(
                topics=topic_words,
                texts=tokens,
                corpus=corpus,
                dictionary=dictionary,
                coherence='c_uci'
            )

            coherence_cnpmi = CoherenceModel(
                topics=topic_words,
                texts=tokens,
                corpus=corpus,
                dictionary=dictionary,
                coherence='c_npmi'
            )

            coherence_cv = coherence_cv.get_coherence()
            wandb.log({'Coherence CV': coherence_cv})
            wandb.log({'Coherence UMASS': coherence_umass.get_coherence()})
            wandb.log({'Coherence UCI': coherence_cuci.get_coherence()})
            wandb.log({'Coherence NPMI': coherence_cnpmi.get_coherence()})
            number_topics = topic_model.get_topic_info().shape[0] - 1
            wandb.log({'Topic Number': number_topics})
            wandb.log(
                {'Uncategorized Post Number': topic_model.get_topic_info().at[0, 'Count']})

            model_name = f'{config_sweep["name"]}_{run.id}'
            topic_model.save(os.path.join(self.path_model, model_name))

    def sweep(self):
        wandb.login()
        sweep_id = wandb.sweep(config_sweep, project=wandb_project)
        wandb.agent(sweep_id, function=self.__train)
```

**Comment by MaartenGr** [comment]
This relates to how fine-grained you want your topics to be. Have you tried inspecting the documents in the topics? That should give you a basic idea of why they were put into separate topics. 

Another trick that you can do is by increasing the `min_topic_size` (or `min_cluster_size` in HDBSCAN). This tends to generate fewer topics which generally removes some of the micro-topics. Moreover, you could use `nr_topics="auto"` to automatically combine some of the topics that might be relevant to one another. Lastly, you could also use hierarchical topic modeling to merge topics that have a maximum distance to one another. This allows you to control when topics should be merged.

**Comment by MaartenGr** [comment]
Just a few things more to note. I see you are using manhattan distance in UMAP. Have you tried "cosine similarity" here? It generally works quite a bit better for embedding-based data. 

Based on your topics, it seems that you are using data related to code/issues/errors. It might be worthwhile to then also choose an embedding model that can represent that accurately, which I am not sure `all-mpnet-base-v2` does. So picking a model trained on embedding code snippets would be worthwhile. 

It might be interesting to use TF-IDF as your input embeddings without `KeyBERTInspired` to show you the effect of an embedding model tuned to your specific data. The benefit of TF-IDF is that it handles whatever domain-specific words you throw at it.

**Comment by zhimin-z** [comment]
> nr_topics="auto"

Is that used like so? I've checked the documentation but it does not say how to assign topics to all docs afterward.
```
    topic_model = BERTopic.load(os.path.join(path_model, name))
    topic_model.reduce_topics(docs, nr_topics='auto')
    topics, probs = topic_model.transform(docs)
```

**Comment by zhimin-z** [comment]
> KeyBERTInspired

BTW, is there any special advantage brought by `KeyBERTInspired` besides training speed? Does it increase coherence score or diversity?

**Comment by zhimin-z** [comment]
> This relates to how fine-grained you want your topics to be. Have you tried inspecting the documents in the topics? That should give you a basic idea of why they were put into separate topics.
> 
> Another trick that you can do is by increasing the `min_topic_size` (or `min_cluster_size` in HDBSCAN). This tends to generate fewer topics which generally removes some of the micro-topics. Moreover, you could use `nr_topics="auto"` to automatically combine some of the topics that might be relevant to one another. Lastly, you could also use hierarchical topic modeling to merge topics that have a maximum distance to one another. This allows you to control when topics should be merged.

It seems that after shrinking the topics, the soft-clustering does not work anymore due to reduced topics. Any suggestion?
![image](https://github.com/MaartenGr/BERTopic/assets/8592144/cc6e6254-23ff-4399-a396-4f071ebbe164)

**Comment by MaartenGr** [comment]
> BTW, is there any special advantage brought by KeyBERTInspired besides training speed? Does it increase coherence score or diversity?

This does not speed up training but its main benefit is that it generally creates topic representations that are a bit more coherent to most people. This does not necessarily mean that the coherence or diversity scores will increase, which is generally not the best metric to pursue. 

> It seems that after shrinking the topics, the soft-clustering does not work anymore due to reduced topics. Is that due to the model loading from saved models rather than training from scratch? Any suggestion?

It is difficult to say without seeing the full code. For example, it is not clear to me how you saved the model (safetensors/pytorch/pickle), which version of BERTopic you are using, or what the training code is. 

Based on your error, my guess would be that `docs`  and `topic_model.topics_` are not of the same size.
